controllers/reconcilers/kserve_isvc_gateway_reconciler.go

```diff
@@ -0,0 +1,279 @@
+
+	istiov1beta1 "istio.io/api/networking/v1beta1"
+	istioclientv1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+)
+
+var _ SubResourceReconciler = (*KserveGatewayReconciler)(nil)
+var meshNamespace string
+
+type KserveGatewayReconciler struct {
+	client         client.Client
+	clientReader   client.Reader
+	secretHandler  resources.SecretHandler
+	gatewayHandler resources.GatewayHandler
+	deltaProcessor processors.DeltaProcessor
+}
+// The clientReader uses the API server to retrieve Secrets that are not cached. By default, only Secrets with the specific label "opendatahub.io/managed: true" are cached.
+func NewKserveGatewayReconciler(client client.Client, clientReader client.Reader) *KserveGatewayReconciler {
+
+	return &KserveGatewayReconciler{
+		client:         client,
+		clientReader:   clientReader,
+		secretHandler:  resources.NewSecretHandler(client),
+		gatewayHandler: resources.NewGatewayHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveGatewayReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling KServe local gateway for Kserve InferenceService")
+
+	_, meshNamespace = utils.GetIstioControlPlaneName(ctx, r.client)
+
+	// return if Address.URL is not set
+	if isvc.Status.Address != nil && isvc.Status.Address.URL == nil {
+		log.V(1).Info("Waiting for the URL as the InferenceService is not ready yet")
+		return nil
+	}
+
+	// return if serving cert secret in the source namespace is not created
+	srcCertSecret := &corev1.Secret{}
+	err := r.clientReader.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace}, srcCertSecret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.V(1).Info(fmt.Sprintf("Waiting for the creation of the serving certificate Secret(%s) in %s namespace", isvc.Name, isvc.Namespace))
+			return nil
+		}
+		return err
+	}
+
+	// Copy src secret to destination namespace when there is not the synced secret.
+	// This use clientReader because the secret that it looks for is not cached.
+	copiedCertSecret := &corev1.Secret{}
+	err = r.clientReader.Get(ctx, types.NamespacedName{Name: fmt.Sprintf("%s-%s", isvc.Name, isvc.Namespace), Namespace: meshNamespace}, copiedCertSecret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if err := r.copyServingCertSecretFromIsvcNamespace(ctx, srcCertSecret, nil); err != nil {
+				log.V(1).Error(err, fmt.Sprintf("Failed to copy the serving certificate Secret(%s) to %s namespace", srcCertSecret.Name, meshNamespace))
+				return err
+			}
+		}
+		return err
+	}
+
+	// Recreate copied secrt when src secret is updated
+	if !reflect.DeepEqual(srcCertSecret.Data, copiedCertSecret.Data) {
+		log.V(1).Info(fmt.Sprintf("Recreating for serving certificate Secret(%s) in %s namespace", copiedCertSecret.Name, meshNamespace))
+		if err := r.copyServingCertSecretFromIsvcNamespace(ctx, srcCertSecret, copiedCertSecret); err != nil {
```

Comment:
Without the `Watch` on the controller, I'm failing to understand how this will catch the certificate rotation, because looks like none of the watched & owned resources will be updated and a reconcile won't be triggered.

I think you should add [these lines](https://github.com/opendatahub-io/odh-model-controller/blob/86cc36042347e1817f7263aacc1ceb818468e85b/main.go#L127-L145), [over here](https://github.com/opendatahub-io/odh-model-controller/blob/86cc36042347e1817f7263aacc1ceb818468e85b/controllers/suite_test.go#L146) so that units can catch this (in case there is an issue).

Reply:
As I commented from [here](https://github.com/opendatahub-io/odh-model-controller/pull/221#discussion_r1656422292), watch can not catch the Secret because it does not have the specific level. I need to think more to solve this rotation part but at the moment, Watch is not the right approach.